src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -191,17 +207,44 @@ double LMNNFunction<MetricType>::Evaluate(const arma::mat& transformation)
   // Apply metric over dataset.
   transformedDataset = transformation * dataset;
 
-  // Calculate norm of change in transformation.
   double transformationDiff = 0;
-  if (!transformationOld.is_empty())
+  if (!transformationOld.is_empty() && iteration++ % range == 0)
   {
+    // Calculate norm of change in transformation.
     transformationDiff = arma::norm(transformation - transformationOld);
-  }
 
-  if (iteration++ % range == 0)
+    if (impBounds)
+    {
+      arma::uvec points(dataset.n_cols);
+
+      // Track number of data points to use for impostors calculatiom.
+      size_t numPoints = 0;
+
+      for (size_t i = 0; i < dataset.n_cols; i++)
+      {
+        if (transformationDiff * (2 * norm(i) + norm(impostors(k - 1, i)) +
+            norm(impostors(k, i))) > distance(k, i) -  distance(k - 1, i))
+        {
+          points(numPoints++) = i;
+        }
+      }
+
+      points.resize(numPoints);
+
+      // Re-calculate impostors on transformed dataset.
+      constraint.Impostors(impostors, distance,
+          transformedDataset, labels, points);
+    }
+    else
+    {
+      // Re-calculate impostors on transformed dataset.
+      constraint.Impostors(impostors, distance, transformedDataset, labels);
+    }
+  }
+  else if (iteration++ % range == 0)
   {
-    // Re-calculate impostors on transformed dataset.
-    constraint.Impostors(impostors, distance, transformedDataset, labels);
+      // Re-calculate impostors on transformed dataset.
+      constraint.Impostors(impostors, distance, transformedDataset, labels);
```

Comment:
I think the extra indent wasn't necessary here.

Reply:
Done!